src/main/scala/org/scalameter/persistence/codecs/complex.scala

```diff
@@ -0,0 +1,116 @@
+package org.scalameter.persistence.codecs
+
+import java.util.Date
+import jawn.ast.{JArray, JNull, JObject, JValue}
+import scala.collection.immutable
+import scala.util.Try
+import org.scalameter._
+
+
+object OptionCodec extends JSONCodec[Option[_]] {
+  def encode(obj: Option[_]): JValue = obj match {
+    case Some(v) =>
+      JObject.fromSeq(Seq(
+        "_value" -> Codecs.encode(v),
+        "_type" -> StringCodec.encode(v.getClass.getName)
+      ))
+    case None => JNull
+  }
+
+  def decode(jV: JValue)(implicit hint: TypeHint): Try[Option[_]] = Try {
+    jV match {
+      case JNull => None
+      case jObj: JObject =>
+        implicit val hint = TypeHint(Class.forName(StringCodec.unsafeDecode(jObj.get("_type"))))
+        Some(Codecs.unsafeDecode(jObj.get("_value")))
+      case _ => sys.error(s"Expected JObject in OptionCodec. Got: $jV.")
+    }
+  }
+}
+
+object SeqCodec extends JSONCodec[Seq[_]] {
+  def encode(obj: Seq[_]): JValue = JArray(obj.map { v =>
+    JObject.fromSeq(Seq(
+      "_value" -> Codecs.encode(v),
+      "_type" -> StringCodec.encode(v.getClass.getName)
+    ))
+  }(collection.breakOut))
+
+  def decode(jV: JValue)(implicit hint: TypeHint): Try[Seq[_]] = Try {
+    jV match {
+      case JArray(vs) => vs.map {
+        case jObj: JObject =>
+          implicit val hint = TypeHint(Class.forName(StringCodec.unsafeDecode(jObj.get("_type"))))
+          Codecs.unsafeDecode(jObj.get("_value"))
+        case v => sys.error(s"Expected JObject as array element in SeqCodec. Got: $v.")
+      }(collection.breakOut)
+      case _ => sys.error(s"Expected JArray in SeqCodec. Got: $jV.")
+    }
+  }
+}
+
+object KeyMapCodec extends JSONCodec[Map[Key[_], Any]] {
+  def encode(obj: Map[Key[_], Any]): JValue = JArray.fromSeq(obj.map { case (k, v) =>
+    KeyValueCodec.encode(k.fullname -> v)
```

Comment:
@axel22: Saving of type hints here can be avoided, but it would require to enforce TypeTag upon key creation, like this:
`class Key[T: scala.reflect.runtime.universe.TypeTag]`
and to add type reconstruction based on it.
Do you see any contraindications?